### ISSUE TITLE
`Literal` parameter rules not strictly enforced

### ISSUE THREAD

**Issue by EtaoinWu** [microsoft/pyright#9964]
**Describe the bug**
Pyright does not strictly enforce the rules for `Literal` parameters until the `Literal` annotation is used "in a context that expects a type".

**Code or Screenshots**
In the following code:

```python
x = object()
y = object()
XorY = Union[Literal[x], Literal[y]]
var: XorY = x
XorY2 = Literal[x] | Literal[y]
```

Pyright does not complain on the third line, but complains about the last line. The error on the last line is also pretty cryptic: `Operator "|" not supported for types "Literal" and "Literal"`.

**VS Code extension or command-line**
VSCode, but reproduced on the online pyright playground at 1.1.394 with python 3.13.

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug.

`Literal` is a special form that is intended to be used in type expressions. You're using it here in a value expression. The typing spec does not indicate how type checkers should evaluate the types of special forms used in value expressions. And indeed, the runtime type for most special forms is an undocumented class that is not defined in typeshed stubs.

If you use `Literal` in a type expression, the typing spec dictates what forms are legal and illegal for this special form. See [this section](https://typing.readthedocs.io/en/latest/spec/literal.html#legal-and-illegal-parameterizations) for details.

It's not clear to me what you're attempting to do with this code. If you provide more details, I might be able to suggest a better approach.
